rust/namespace/src/lib.rs

```diff
@@ -233,59 +235,118 @@ pub mod namespace {
             .checked_add(1)
             .ok_or(ErrorCode::NumericalOverflowError)?;
 
-        let accounts = ItemClassCacheNamespace {
-            item_class: ctx.accounts.artifact.to_account_info(),
-            namespace: ctx.accounts.namespace.to_account_info(),
-            instructions: ctx.accounts.instructions.to_account_info(),
-        };
+        let rd_program = ctx.accounts.raindrops_program.to_account_info();
 
-        let cpi_ctx = CpiContext::new(ctx.accounts.item_program.to_account_info(), accounts);
+        if raindrops_item::check_id(&rd_program.key()) {
+            let accounts = ItemClassCacheNamespace {
+                item_class: ctx.accounts.artifact.to_account_info(),
+                namespace: namespace.to_account_info(),
+                instructions: ctx.accounts.instructions.to_account_info(),
+            };
+
+            item_class_cache_namespace(CpiContext::new(rd_program, accounts), args.page)
+        } else if raindrops_matches::check_id(&rd_program.key()) {
+            let accounts = MatchCacheNamespace {
+                match_instance: ctx.accounts.artifact.to_account_info(),
+                namespace: namespace.to_account_info(),
+                instructions: ctx.accounts.instructions.to_account_info(),
+            };
 
-        item_class_cache_namespace(cpi_ctx, index.page)
+            match_cache_namespace(CpiContext::new(rd_program, accounts), args.page)
+        } else if crate::id().eq(&rd_program.key()) {
+            let artifact_ns = &mut Account::<'_, Namespace>::try_from(&ctx.accounts.artifact)?;
+
+            let mut cached = false;
+            let mut new_namespaces: Vec<NamespaceAndIndex> = vec![];
+            for ns in artifact_ns.namespaces.clone().unwrap() {
+                if ns.namespace == namespace.key() && !cached {
+                    cached = true;
+                    new_namespaces.push(NamespaceAndIndex {
```

Comment:
Why did you switch to creating a new struct in this call, vs. muting the original like you did in the previous files above?

Reply:
switched to this way, more clear imo

Reply:
If you're going to switch, you probably should switch them all then.